File: systems/Control/PIDControllerBase.cpp
#include "PIDControllerBase.h"

namespace gnc
{
    PIDControllerBase::PIDControllerBase(float kp, float ki, float kd, float outMin, float outMax)
        : m_Kp(kp), m_Ki(ki), m_Kd(kd), m_OutMin(outMin), m_OutMax(outMax),
          m_Integral(0.0f), m_PrevMeasurement(0.0f), m_Initialized(false)
    {
    }

    PIDControllerBase::SaturationResult PIDControllerBase::saturate(float value, float outMin, float outMax)
    {
        if (value >= outMax)
        {
            return {outMax, /*atUpper=*/true, /*atLower=*/false};
        }
        if (value <= outMin)
        {
            return {outMin, /*atUpper=*/false, /*atLower=*/true};
        }
        return {value, /*atUpper=*/false, /*atLower=*/false};
    }

    float PIDControllerBase::update(float setpoint, float measurement, float dt)
    {
        float error = setpoint - measurement;

        // Tentative integral (Riemann sum) -- only committed below if this
        // step isn't saturated in the direction the error is pushing.
        float tentativeIntegral = m_Integral + error * dt;

        // D-on-measurement, suppressed on the very first call: there's no
        // prevMeasurement yet to difference against.
        float derivative = 0.0f;
        if (m_Initialized)
        {
            derivative = -(measurement - m_PrevMeasurement) / dt;
        }

        float rawOutput = m_Kp * error + m_Ki * tentativeIntegral + m_Kd * derivative;

        SaturationResult saturation = saturate(rawOutput, m_OutMin, m_OutMax);
        float output = saturation.value;

        // Conditional integration: freeze the integral only when saturated
        // AND the error is still pushing further into that same bound.
        bool pushingIntoSaturation = (saturation.atUpper && error > 0.0f) ||
                                     (saturation.atLower && error < 0.0f);
        if (!pushingIntoSaturation)
        {
            m_Integral = tentativeIntegral;
        }

        m_PrevMeasurement = measurement;
        m_Initialized = true;

        return output;
    }

    void PIDControllerBase::reset()
    {
        m_Integral = 0.0f;
        m_PrevMeasurement = 0.0f;
        m_Initialized = false;
    }

    float PIDControllerBase::kp() const { return m_Kp; }
    float PIDControllerBase::ki() const { return m_Ki; }
    float PIDControllerBase::kd() const { return m_Kd; }
    float PIDControllerBase::outMin() const { return m_OutMin; }
    float PIDControllerBase::outMax() const { return m_OutMax; }
}

File: MATLAB Design and Experimentation/quadrotor gnc 2026/src/PIDControllerBase.m
classdef PIDControllerBase < handle
    properties
        Kp, Ki, Kd
        dt
        outMin, outMax
        integral
        prevMeasurement
        initialized
    end

    methods
        function obj = PIDControllerBase(Kp, Ki, Kd, dt, outMin, outMax)
            obj.Kp = Kp;
            obj.Ki = Ki;
            obj.Kd = Kd;
            obj.dt = dt;
            obj.outMin = outMin;
            obj.outMax = outMax;
            obj.integral = 0;
            obj.prevMeasurement = 0;
            obj.initialized = false;
        end

        function out = update(obj, setpoint, measurement)
            error = setpoint - measurement;
            
            Pterm = obj.Kp * error;
            
            integral_candidate = obj.integral + error*obj.dt;
            Iterm = obj.Ki * integral_candidate;
            
            if (obj.initialized == 0)
                Dterm = 0;
            else
                Dterm = -obj.Kd * (measurement - obj.prevMeasurement) / obj.dt;
            end

            out_unsat = Pterm + Iterm + Dterm;
            [out, atUpper, atLower] = obj.saturate(out_unsat, obj.outMin, obj.outMax);
            freeze = (atUpper && error > 0) || (atLower && error < 0);
            if ~freeze
                obj.integral = integral_candidate;
            end

            obj.prevMeasurement = measurement; % Update previous measurement
            obj.initialized = true; % Set initialization flag
        end

        function reset(obj)
            obj.integral = 0; % Reset integral term
            obj.prevMeasurement = 0; % Reset previous measurement
            obj.initialized = false; % Reset initialization flag
        end
    end

    methods (Static)
        function [out, atUpper, atLower] = saturate(out_unsat, outMin, outMax)
            out = max(outMin, min(out_unsat, outMax));
            atUpper = (out >= outMax);
            atLower = (out <= outMin);
        end
    end
end